org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ILanguageServerAccess.java

```diff
@@ -0,0 +1,158 @@
+
+/**
+ * API for interacting with a running language server.
+ * 
+ * @author Sven Efftinge - Initial contribution and API
+ */
+public interface ILanguageServerAccess {
+	class Context {
+		private final Resource resource;
+
+		private final Document document;
+
+		private final boolean isDocumentOpen;
+
+		private final CancelIndicator cancelChecker;
+
+		public Context(Resource resource, Document document, boolean isDocumentOpen, CancelIndicator cancelChecker) {
+			this.resource = resource;
+			this.document = document;
+			this.isDocumentOpen = isDocumentOpen;
+			this.cancelChecker = cancelChecker;
+		}
+
+		@Override
+		public String toString() {
+			ToStringBuilder b = new ToStringBuilder(this);
+			b.add("resource", resource);
+			b.add("document", document);
+			b.add("isDocumentOpen", isDocumentOpen);
+			b.add("cancelChecker", cancelChecker);
+			return b.toString();
+		}
+
+		public Resource getResource() {
+			return resource;
+		}
+
+		public Document getDocument() {
+			return document;
+		}
+
+		public boolean isDocumentOpen() {
+			return isDocumentOpen;
+		}
+
+		public CancelIndicator getCancelChecker() {
+			return cancelChecker;
+		}
+	}
+
+	class IndexContext {
+		private final IResourceDescriptions index;
+
+		private final CancelIndicator cancelChecker;
+
+		public IndexContext(IResourceDescriptions index, CancelIndicator cancelChecker) {
+			this.index = index;
+			this.cancelChecker = cancelChecker;
+		}
+
+		@Override
+		public String toString() {
+			ToStringBuilder b = new ToStringBuilder(this);
+			b.add("index", index);
+			b.add("cancelChecker", cancelChecker);
+			return b.toString();
+		}
+
+		public IResourceDescriptions getIndex() {
+			return index;
+		}
+
+		public CancelIndicator getCancelChecker() {
+			return cancelChecker;
+		}
+	}
+
+	interface IBuildListener {
+		void afterBuild(List<IResourceDescription.Delta> deltas);
+	}
+
+	/**
+	 * provides read access to a fully resolved resource and Document.
+	 */
+	<T extends Object> CompletableFuture<T> doRead(String uri, Function<ILanguageServerAccess.Context, T> function);
+
+	/**
+	 * Provides read access to the Xtext index.
+	 * 
+	 * @since 2.18
+	 */
+	<T extends Object> CompletableFuture<T> doReadIndex(
+			Function<? super ILanguageServerAccess.IndexContext, ? extends T> function);
+
+	/**
+	 * registers a build listener on the this language server
+	 */
+	void addBuildListener(ILanguageServerAccess.IBuildListener listener);
+
+	/**
+	 * @return the language client facade. It usually also implements Endpoint, which can be used to call non-standard
+	 *         extensions to the LSP.
+	 */
+	LanguageClient getLanguageClient();
+
+	/**
+	 * Creates a new {@link ResourceSet} to be used with the {@link IChangeSerializer} API.
+	 * 
+	 * The {@link IChangeSerializer} allows to calculate text edits from model changes. In order not to mess up the
+	 * originals, the resp. models should be loaded into a new resource set which this method provides.
```

Comment:
Not very useful to have abbreviations here
resp = respective?
resp = responsive?
resp = RESt in Pepperonis?